File: datarot.py
#!/usr/bin/python3
import os
import sys
import math

SMALL = 0.001
m = 0.285
c = -32.477


def FindIntersection(m1, c1, m2, c2):
   """
      Takes the equations for 2 lines and finds the intersection
      Returns the intersection (x,y) and a Boolean flag to indicate
      that they did indeed intersect

      Line 1: y=m1*x + c1
      Line 2: y=m2*x + c2
      
      When y matches,
      m1*x + c1 = m2*x + c2
      So...
      m1*x - m2*x = c2 - c1
      x(m1-m2) = c2 - c1
      x = (c2-c1)/(m1-m2)
      
      and
      y = m1*x + c1
   """

   if(abs(m1-m2) < SMALL):
      return(0.0, 0.0, False)

   x = (c2-c1)/(m1-m2)
   y = m1*x + c1
   
   return(x, y, True)


def TranslateDataArrays(xData, yData, ndata, transX, transY):
   """
      Takes x and y data arrays, the number of items and an x and y
      translation. Applies the translation to each data point and
      returns the new arrays
   """
   for i in range(ndata):
      xData[i] = xData[i] + transX
      yData[i] = yData[i] + transY
   return(xData, yData)


def FindAngle(m1, m2):
   """ 
      Finds the angle between two lines based on their slopes.
      Returns the angle (0.0 if the slopes are the same)

      $ tan(\theta) = (m1-m2)/(1 + m1*m2) $ 
   """
   if(abs(m1-m2) < SMALL):
      return(0.0)

   tanTheta = (m1-m2)/(1 + m1*m2)
   return(math.atan(tanTheta))


def RotateDataArray(xData, yData, nData, theta):
   """
      Takes two arrays of (2D) data points, the number of points and an angle.
      Rotates each data 2D point about the origin by the specified angle
   """
   
   for i in range(nData):
      x = xData[i] * math.cos(theta) - yData[i] * math.sin(theta)
      y = xData[i] * math.sin(theta) + yData[i] * math.cos(theta)
      xData[i] = x
      yData[i] = y
   return(xData, yData)

if(__name__ == '__main__'):
   # Main program
   with open ("test/Everything_NR2_SklearnGBReg.csv") as file:
      ndata = 0
      xData = []
      xDataOrig = []
      yData = []
    
      for buffer in file:
         if(buffer[0] != '#'):
            fields = buffer.split(',')
            x = float(fields[1])
            y = float(fields[2])
            xData.append(x)
            xDataOrig.append(x)
            yData.append(y)
            ndata = ndata+1

      print("Original Data:")
      
      for i in range(ndata):
         print("%f,%f" % (xData[i], yData[i]))

      intersectX, intersectY, ok = FindIntersection(m, c, 1, 0)
      if(ok):
         xData, yData = TranslateDataArrays(xData, yData, ndata,
                                            -intersectX, -intersectY)
         angle = FindAngle(m, 1)
         xData, yData = RotateDataArray(xData, yData, ndata, -angle)
         xData, yData = TranslateDataArrays(xData, yData, ndata,
                                               intersectX, intersectY)

      print("Int: %.3f %.3f" % (intersectX, intersectY))
      print("Ang: %.3f" % (180.0 * angle / math.pi))

      print("Rotated Data:")
      for i in range(ndata):
         print("%f,%f" % (xData[i], yData[i]))

      print("Rotated Y Data (X original):")
      for i in range(ndata):
         print("%f,%f" % (xDataOrig[i], yData[i]))

File: datarot.c
#define TEST
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "svdfit.h"
#include "bioplib/array.h"


#define SMALL 0.001
#define MAXDATA 5000
#define MAXBUFF 160

int main(int argc, char **argv);
BOOL FindIntersection(REAL m1, REAL c1, REAL m2, REAL c2,
		      REAL *intersectX, REAL *intersectY);
void TranslateData(REAL *xData, REAL *yData, int ndata, REAL transX, REAL transY);
REAL FindAngle(REAL m1, REAL m2);
void RotateData(REAL *xData, REAL *yData, int ndata, REAL angle);








int main(int argc, char **argv)
{
   FILE *fp = NULL;
   
   if((fp=fopen("corrected.csv", "r"))!=NULL)
   {
      int ndata = 0, i;
      REAL xData[MAXDATA],
         yData[MAXDATA],
         m = 0.285,
         c = -32.477,
         intersectX,
         intersectY;
      
#ifdef TEST
      ndata = 0;
      
      for(i=-60; i<-30; i+=5)
      {
         xData[ndata] = i;
         /*yData[ndata] = i;*/
	 yData[ndata] = 2 * i + 5;
         ndata++;
      }
      printf("Original Data\n");
      
      for (i=0; i<ndata; i++)
      {
         printf("%f,%f\n", xData[i], yData[i]);
      }
#else
      char buffer[MAXBUFF];
      while(fgets(buffer, MAXBUFF, fp))
      {
         REAL x, y;
         
         sscanf(buffer, "%lf,%lf", &x, &y);
         xData[ndata] = x;
         yData[ndata] = y;
         ndata++;
      }
#endif
      
      if(FindIntersection(m, c, 1, 0, &intersectX, &intersectY))
      {
         REAL angle;
         
         TranslateData(xData, yData, ndata, -intersectX, -intersectY);
         angle = FindAngle(m, 1);
         RotateData(xData, yData, ndata, angle);
         TranslateData(xData, yData, ndata, intersectX, intersectY);
#ifdef TEST
         printf("Int: %.3f %.3f\n", intersectX, intersectY);
         printf("Ang: %.3f\n", 180*angle/PI);
#endif
         
      }
      for (i=0; i<ndata; i++)
      {
         printf("%f,%f\n", xData[i], yData[i]);
      }
   }
   return(0);
}


BOOL FindIntersection(REAL m1, REAL c1, REAL m2, REAL c2,
		      REAL *intersectX, REAL *intersectY)
{
   /* Line 1: y=m1*x + c1
      Line 2: y=m2*x + c2
      
      When y matches,
      m1*x + c1 = m2*x + c2
      So...
      m1*x - m2*x = c2 - c1
      x(m1-m2) = c2 - c1
      x = (c2-c1)/(m1-m2)
      
      and
      y = m1*x + c1;
   */
   REAL x, y;

   if(ABS(m1-m2) < SMALL)
   {
      return(FALSE);
   }
   
   x = (c2-c1)/(m1-m2);
   y = m1*x + c1;
   
   *intersectX = x;
   *intersectY = y;
   return(TRUE);
}

void TranslateData(REAL *xData, REAL *yData, int ndata, REAL transX, REAL transY)
{
   int i;
   for(i=0; i<ndata; i++)
   {
      xData[i] += transX;
      yData[i] += transY;
   }
}

REAL FindAngle(REAL m1, REAL m2)
{
   REAL tanTheta;
   
   /* $ tan(\theta) = (m1-m2)/(1 + m1*m2) $ */
   if(ABS(m1-m2) < SMALL)
   {
      return(0.0);
   }

   tanTheta = (m1-m2)/(1 + m1*m2);
   return(atan(tanTheta));
}

void RotateData(REAL *xData, REAL *yData, int nData, REAL theta)
{
   int i;
   for(i=0; i<nData; i++)
   {
      REAL x, y;
      x = xData[i] * cos(theta) - yData[i] * sin(theta);
      y = xData[i] * sin(theta) + yData[i] * cos(theta);
      xData[i] = x;
      yData[i] = y;
   }
}

